File: src/classes/containers/cubeContainer.cpp
#include "cubeContainer.hpp"
#include <glm/glm.hpp>

CubeContainer::CubeContainer(glm::vec3 position, glm::vec3 size, bool forcedInside) : Container(position, forcedInside) {
    this->size = size;
}

glm::vec3 CubeContainer::getSize() {
    return size;
}

void CubeContainer::setSize(glm::vec3 size) {
    this->size = size;
}

bool Container::getForcedInside() {
    return forcedInside;
}

void CubeContainer::collideWith(Sphere* sphere) {
    // Collision resolution code
    glm::vec3 spherePosition = sphere->position;
    float sphereRadius = sphere->radius;
    glm::vec3 min = position - size / 2.0f; // getting the minimum points of the container
    glm::vec3 max = position + size / 2.0f; // getting the maximum points of the container
    // Check if the sphere is inside the container
    if (!forcedInside && !(min.x <= spherePosition.x && spherePosition.x <= max.x && min.y <= spherePosition.y && spherePosition.y <= max.y && min.z <= spherePosition.z && spherePosition.z <= max.z)) {
        // the sphere is outside the container
        return;
    }

    // Collision resolution code
    if (spherePosition.x - sphereRadius < min.x) {
        sphere->position.x = min.x + sphereRadius;
    } 
    if (spherePosition.x + sphereRadius > max.x) {
        sphere->position.x = max.x - sphereRadius;
    }
    if (spherePosition.y - sphereRadius < min.y) {
        sphere->position.y = min.y + sphereRadius;
    }
    if (spherePosition.y + sphereRadius > max.y) {
        sphere->position.y = max.y - sphereRadius;
    }
    if (spherePosition.z - sphereRadius < min.z) {
        sphere->position.z = min.z + sphereRadius;
    }
    if (spherePosition.z + sphereRadius > max.z) {
        sphere->position.z = max.z - sphereRadius;
    }
}

File: rust/src/classes/containers/cube_container.rs
use nalgebra::{Vector3};

use crate::classes::particles::sphere::Sphere;
use crate::classes::container::{ContainerTrait, Container};

pub struct CubeContainer {
    container: Container,
    size: Vector3<f32>,
}

impl CubeContainer {
    pub fn new(position: Vector3<f32>, size: Vector3<f32>, forced_inside: bool) -> Self {
        CubeContainer {
            container: Container::new(position, forced_inside),
            size,
        }
    }
}

impl ContainerTrait for CubeContainer {
    fn get_size(&self) -> Vector3<f32> {
        self.size
    }

    fn get_position(&self) -> Vector3<f32> {
        self.container.position
    }

    fn set_position(&mut self, position: Vector3<f32>) {
        self.container.position = position;
    }

    fn set_size(&mut self, size: Vector3<f32>) {
        self.size = size;
    }

    fn collide_with(&self, sphere: &mut Sphere) {
        let sphere_position = sphere.particle.position;
        let sphere_radius = sphere.radius;
        let min = self.container.position - self.size / 2.0;
        let max = self.container.position + self.size / 2.0;

        if !self.container.forced_inside && !(min.x <= sphere_position.x && sphere_position.x <= max.x && min.y <= sphere_position.y && sphere_position.y <= max.y && min.z <= sphere_position.z && sphere_position.z <= max.z) {
            return;
        }

        if sphere_position.x - sphere_radius < min.x {
            sphere.particle.position.x = min.x + sphere_radius;
        }
        if sphere_position.x + sphere_radius > max.x {
            sphere.particle.position.x = max.x - sphere_radius;
        }
        if sphere_position.y - sphere_radius < min.y {
            sphere.particle.position.y = min.y + sphere_radius;
        }
        if sphere_position.y + sphere_radius > max.y {
            sphere.particle.position.y = max.y - sphere_radius;
        }
        if sphere_position.z - sphere_radius < min.z {
            sphere.particle.position.z = min.z + sphere_radius;
        }
        if sphere_position.z + sphere_radius > max.z {
            sphere.particle.position.z = max.z - sphere_radius;
        }
    }
}